### PR TITLE
Add axis argument to multidim

### PR DESCRIPTION
**Pull request by kwohlfahrt**
This adds an optional axis argument to the functions in the `multidim` module. Probably of most interest to @aaren, what do you think?

Again, built on top of #129, so merge that first.

### CODE REVIEW

**Comment by aaren** [comment]
Ace. I'll review when I've a chance.

**Comment by grlee77** [comment]
This is nice and likely to be very useful.  

Why did you decide on `axis` as opposed to `axes` for the argument name?  I think the second makes it clearer that a sequence as opposed to an integer is needed as input.

I would prefer `axes` for these and `axis` for the 1D cases only, but could be convinced otherwise.  There is precedent for the name `axes` for functions in Numpy as well in the 2D and nD FFT routines.  `np.transpose` is another example.

**Comment by kwohlfahrt** [comment]
I originally allowed an integer to be used as well as a sequence, and wanted `axis` to match with functions like `numpy.maximum`. But now that it is changed `axes is probably more appropriate.

**Comment by grlee77** [comment]
This looks good to me. 

In terms of clarity to users, I am not sure if it makes sense to both allow repeats of the same axis and restrict the axes list to length 2 for `dwt2`/`idwt2`.  What if the user wanted to do axes = (1, 1, 2)?  There are still only 2 unique axes so is it a 2D transform?  seems ambiguous to me, so I thought maybe it was clearer to only allow such cases in the general `dwtn`/`idwtn`.  I am okay if you want to leave it as is though.

**Comment by kwohlfahrt** [comment]
The reason I went for needing exactly two axes is to keep the current output format `LL, (HL, LH, HH)`. This is only possible if exactly two axes are specified. We can require them to be unique, but I don't see a particular reason for that, so we may as well be as flexible as possible.

File: pywt/_multidim.py [hunk 1]
```diff
@@ -237,7 +155,11 @@ def dwtn(data, wavelet, mode='symmetric'):
         raise ValueError("Input data must be at least 1D")
     coeffs = [('', data)]
 
-    for axis in range(data.ndim):
+    if axes is None:
+        axes = range(data.ndim)
+    axes = (a + data.ndim if a < 0 else a for a in axes)
+
+    for axis in sorted(axes):
```
Comment (aaren) [inline pywt/_multidim.py]: I think the use of sorted axes needs documenting. In `np.fft.fftn` the axes are transformed over in the order given.

File: pywt/_multidim.py [hunk 2]
```diff
@@ -294,9 +220,13 @@ def idwtn(coeffs, wavelet, mode='symmetric'):
     if any(s != coeff_shape for s in coeff_shapes):
         raise ValueError("`coeffs` must all be of equal size (or None)")
 
-    for axis in reversed(range(dims)):
+    if axes is None:
+        axes = range(ndim)
+    axes = (a + ndim if a < 0 else a for a in axes)
+
+    for key_length, axis in reversed(list(enumerate(sorted(axes)))):
```
Comment (aaren) [inline pywt/_multidim.py]: Similarly here, document sorted axes.

File: pywt/_multidim.py [hunk 3]
```diff
@@ -53,46 +57,17 @@ def dwt2(data, wavelet, mode='symmetric'):
 
     """
     data = np.asarray(data)
-    if data.ndim != 2:
-        raise ValueError("Expected 2-D data array")
-
-    if not isinstance(wavelet, Wavelet):
-        wavelet = Wavelet(wavelet)
+    if axes is None:
+        axes = range(-2, 0)
```
Comment (aaren) [inline pywt/_multidim.py]: How about just setting `axes=(-2, -1)` as the default arg? It is more explicit and is also what `np.fft.fft2` uses.

File: pywt/_multidim.py [hunk 4]
```diff
@@ -31,6 +31,10 @@ def dwt2(data, wavelet, mode='symmetric'):
         Wavelet to use
     mode : str, optional
         Signal extension mode, see Modes (default: 'symmetric')
+    axes : 2-tuple of ints, optional
+        Axes over which to compute the DWT. Repeated elements mean the DWT will
+        be performed multiple times along these axes. A value of `None` (the
```
Comment (aaren) [inline pywt/_multidim.py]: Documentation is at odds with implementation. State the restriction to two axes if that is being used.

File: pywt/_multidim.py [hunk 5]
```diff
@@ -107,6 +82,10 @@ def idwt2(coeffs, wavelet, mode='symmetric'):
         Wavelet to use
     mode : str, optional
         Signal extension mode, see Modes (default: 'symmetric')
+    axes : 2-tuple of ints, optional
+        Axes over which to compute the IDWT. Repeated elements mean the IDWT
+        will be performed multiple times along these axes. A value of `None`
```
Comment (aaren) [inline pywt/_multidim.py]: Again, state forced use of two axes.

File: pywt/_multidim.py [hunk 6]
```diff
@@ -118,84 +97,19 @@ def idwt2(coeffs, wavelet, mode='symmetric'):
...
 
     """
-    if len(coeffs) != 2 or len(coeffs[1]) != 3:
-        raise ValueError("Invalid coeffs param")
-
     # L -low-pass data, H - high-pass data
     LL, (HL, LH, HH) = coeffs
+    if axes is None:
+        axes = range(-2, 0)
+    axes = tuple(axes)
```
Comment (aaren) [inline pywt/_multidim.py]: `axes=(-2, -1)` as default arg is simpler.

**Comment by aaren** [comment]
How to treat the axes is where I got bogged down when implementing this. Tricky points are repeated axes and what to do if you have `m` axes and `n` dimensions and `m!=n`.

File: pywt/_multidim.py (hunk 5 above)
Comment (kwohlfahrt) [inline pywt/_multidim.py]: I thought `2-tuple of ints` was clear that a tuple containing 2 integers is required. I can spell it out again in the body if necessary.

File: pywt/_multidim.py (hunk 3 above)
Comment (kwohlfahrt) [inline pywt/_multidim.py]: Yeah, that is better.

File: pywt/_multidim.py (hunk 1 above)
Comment (kwohlfahrt) [inline pywt/_multidim.py]: Yes, good point. It might work without sorting, will have a look.

File: pywt/_multidim.py (hunk 5 above)
Comment (aaren) [inline pywt/_multidim.py]: Ah yes. Didn't see that. It's ok as-is then.

**Comment by kwohlfahrt** [comment]
I would just treat the dimensionality of the transform as the number of axes it is taken over, rather than the dimensionality of the input.

The repeated axes issue is a bit more interesting. I've chosen to simply ignore it for now. I think the worst case is that the user ends up with *very* different reconstructions, provided they repeat the same axis many times and/or provide axes in different orders. However, that would require deliberate (ab)use of the feature, so I don't think we should worry unless somebody explicitly complains. It is probably a good idea to document best practices for exact reconstruction though.

**Comment by grlee77** [comment]
That sounds reasonable to me.  Please also add the axes arguments to the documentation in 
`/doc/source/ref/2d-dwt-and-idwt.rst`
Corresponding documentation for the nD functions doesn't currently seem to exist.

**Comment by grlee77** [comment]
just a quick heads up:  I added some basic documentation of the nD functions (including the axes arguments) in #138

**Comment by grlee77** [comment]
This looks good Kai, there is just one typo I commented on above.

Don't worry about adding the `axis` argument to the API documentation.  In #139, I am trying to switch these over to being autogenerated as you suggested.

**Comment by kwohlfahrt** [comment]
OK, fixed up now.

**Pull request merged**